cranelift/isle/isle/src/codegen.rs

```diff
@@ -60,7 +68,21 @@ impl<'a, W: Write> BodyContext<'a, W> {
         writeln!(self.out, " {{")
     }
 
-    fn end_block(&mut self, scope: StableSet<BindingId>) -> std::fmt::Result {
+    fn begin_nested(
+        &mut self,
+        nested: Nested<'b>,
+        last_line: &'b str,
+        scope: StableSet<BindingId>,
+    ) -> std::fmt::Result {
+        self.stack.push((nested, last_line, scope));
+        self.begin_block()
+    }
+
+    fn end_nested(&mut self) -> std::fmt::Result {
+        let (_, end, scope) = self.stack.pop().unwrap();
```

Comment:
Can this method take these things as arguments, instead of popping them? IIUC, that would allow the `while let` loop to do `ctx.stack.pop()` instead of `ctx.stack.last_mut()` which would make reasoning about control flow and termination much more straightforward, IMO, since it would all be localized within the same function.

Reply:
Similarly, soft preference for having `begin_nested` (or `begin_block`, and removing `begin_nested`) return the stack entry and having callers do something like

```rust
let entry = ctx.begin_nested()?;
ctx.stack.push(entry);
```

For the same reasons, I think this would make it a little easier to understand what's happening here.